File: tpl/src/rngsse2/examples/examples_SSE_C/mt19937_sse.c

#include<stdio.h>
#include<mt19937.h>

#define NN       100000000UL

int main(void){ 
   long i; unsigned int sum=0;
   mt19937_state state; mt19937_sse_state sse_state;
   mt19937_init_(&state); mt19937_get_sse_state_(&state,&sse_state);
   mt19937_print_state_(&state);
   for(i=0;i<NN;i++) sum+=mt19937_sse_generate_(&sse_state);
   printf("%ld MT19937_SSE pseudorandom numbers generated.\n",NN);
   printf("Fractional part of the total sum of generated numbers: %f\n",sum/4294967296.);
   printf("Next output value: %f\n",mt19937_sse_generate_(&sse_state)/4294967296.);
   return 0;
}

File: examples/examples_SSE_FORTRAN/mt19937_sse.f90

PROGRAM MT19937_Example

   PARAMETER (N=100000000)
   INTEGER i,MT19937_SSE_generate,sum
   REAL val,rsum

   TYPE MT19937_state
     INTEGER z(625)
   END TYPE MT19937_state
   TYPE(MT19937_state) state

   TYPE MT19937_sse_state
     INTEGER z(3770)
   END TYPE MT19937_sse_state
   TYPE(MT19937_sse_state) SSE_state

   CALL MT19937_init ( state )
   CALL MT19937_Get_SSE_state (state,SSE_state)
   CALL MT19937_Print_State (state)

   sum = 0

   DO i=1,N
     sum = sum + MT19937_SSE_generate (SSE_state)
   END DO

   WRITE(*,1) N
   rsum = sum/4294967296.
   val = MT19937_SSE_generate (SSE_state)/4294967296.
   IF (rsum<0) THEN
     rsum=rsum+1
   ENDIF
   IF (val<0) THEN
     val=val+1
   ENDIF
   WRITE(*,2) rsum
   WRITE(*,3) val

1  FORMAT(I9," MT19937_SSE pseudorandom numbers generated.")
2  FORMAT("Fractional part of the total sum of generated numbers: ",F8.6)
3  FORMAT("Next output value: ",F8.6)

END PROGRAM MT19937_Example

